graph/src/data/store/scalar.rs

```diff
@@ -16,6 +20,51 @@ pub type BigDecimal = bigdecimal::BigDecimal;
 #[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub struct BigInt(num_bigint::BigInt);
 
+pub(crate) fn big_decimal_stable_hash(
+    decimal: &BigDecimal,
+    mut sequence_number: impl SequenceNumber,
+    state: &mut impl StableHasher,
+) {
+    // Performance: This does a non-trivial clone under the hood. There should
+    // be a way to get around that.
+    let (int, exp) = decimal.as_bigint_and_exponent();
+    // I tried for some time to come up with a version of this that ended up
+    // preserving backward compatibility when converting between int types. It
+    // turns out this is impossible with the current API. I explored alternative
+    // APIs, but fell short due to really subtle problems. The reason this
+    // doesn't work is that calling sequence_number.next_child() mutates
+    // sequence_number. So, if the &int is negative, when it writes the
+    // is_negative bool it ends up writing that value with a different sequence
+    // number than had we not called next_child. As a result we'll have to
+    // settle with backward compatible changes from any uint type to BigDecimal
+    // and back, but not signed ints.
+    //
+    // My gut feeling is that there is an API that works, and it's somewhere in
+    // the area of having the stable_hash fn yield back some *unused*
+    // SequenceNumber to the caller. Intuitively I think such an API would be
+    // significantly more error prone and will not explore further at this time.
+    exp.stable_hash(sequence_number.next_child(), state);
+    big_int_stable_hash(&int, sequence_number, state);
+}
+
+fn big_int_stable_hash(
+    int: &num_bigint::BigInt,
+    sequence_number: impl SequenceNumber,
+    state: &mut impl StableHasher,
+) {
+    AsInt {
+        is_negative: int.sign() == BigIntSign::Minus,
+        little_endian: &int.to_bytes_le().1,
+    }
+    .stable_hash(sequence_number, state)
+}
```

Comment:
Why not move this into the `stable_hash` implementation below?

And then call `BigInt::stable_hash(&int, sequence_number, state)` in the implementation for `BigDecimal`?

Reply:
Because of the orphan rules.

We can't implement `StableHash` for `bigint::BigInt`, only `data::store::scalar::BigInt`. `big_decimal_stable_hash` only has access to the former and cannot construct a `data::store::scalar::BigInt` without moving out of the `BigDecimal`.

Reply:
Oh, of course. `num_bigint::BigInt` vs. our `sclar::BigInt`. Nevermind.